### ISSUE TITLE
Unloading bug in 3d Tiles branch

### ISSUE THREAD

**Issue by lucasvw** [AnalyticalGraphicsInc/cesium#4673]
Hi everyone,

There seems to be a problem with the unloading of the 3d-tiles branch.

At some camera positions, the same tiles keep on being loaded over and over again (being requested from server).

The following link lands at exactly such a position:

http://demo.virtualcitymap.de/index2.html?startingmap=cesium&lang=en&cameraPosition=13.43017%2C52.57237%2C243.09854&groundPosition=13.43087%2C52.57779%2C46.74016&distance=635.49&pitch=-18.00&heading=4.51&roll=0.02&layerToActivate=buildings&layerToDeactivate=buildings_untextured

I created http://demo.virtualcitymap.de/index2.html which loads the CesiumUnminified version.

In chrome dev tools I am seeing a couple of b3dm files being constantly loaded new, such as 27253.b3dm and 13626.b3dm, which creates a really big amount of network traffic over time :)

Best,
Lucas

**Comment by lilleyse** [comment]
I'm also noticing that in the network tab. Can you create a quick Sandcastle demo for this?

**Comment by jbo023** [comment]
var viewer = new Cesium.Viewer('cesiumContainer');

var tileset = viewer.scene.primitives.add(new Cesium.Cesium3DTileset({
    url : "http://demo.virtualcitymap.de/datasource-data/berlin",
    maximumNumberOfLoadedTiles:3,
    maximumScreenSpaceError:16    
}));

var cameraPosition = new Cesium.Cartesian3(3778492.598335113, 902266.7207133673, 5041956.047869663);
var cam = viewer.scene.camera;
var cameraOptions = {
    "heading": Cesium.Math.toRadians(4.51),
    "pitch": Cesium.Math.toRadians(-18.00),
    "roll": Cesium.Math.toRadians(0.02)
};

var direction = new Cesium.Cartesian3(-0.9323680118559295, -0.14575318580427238, 0.33082608617780795);
var up = new Cesium.Cartesian3(0.318555765161632, 0.10140786762051479, 0.9424641472581052);

cam.setView({
    "destination": cameraPosition,
    "orientation": {
        "direction":direction,
        "up":up
    }
});

The Bug only occurs if we set maximumNumberOfLoadedTiles to a small value.
Jannes

**Comment by lilleyse** [comment]
Is the url still up to date? I'm not noticing any unusual activity in the network tab but the tileset seems wrong.

Captured from a different view:
<img width="770" alt="floating" src="https://cloud.githubusercontent.com/assets/915398/22933743/893bbc04-f29a-11e6-92a6-868c7dc0f5ab.png">

**Comment by jbo023** [comment]
Hi, 

I just pulled the latest 3d-tiles branch and have the same problem. The tileset is not shown correctly. Did something change with the rendering code in the last days? 

But the unloading problem is still there. If i just run the examplecode in a sandcastle, chrome won't stop loading b3dm tiles. But I just noticed that this only happens if you disable the cache. This is really strange.

**Comment by lilleyse** [comment]
I'm looking into why its broken on the latest branch, there should be a fix soon and then I'll get back to this.

**Comment by jbo023** [comment]
I also checked some other datasets, the problem is only with textured datasets.

**Comment by lilleyse** [comment]
There was a recent change in `Model.js` in how program attributes are checked.

Looking at one of the b3dm tiles not all the attributes are specified correctly in `programs` objects:

```json
  "programs": {
    "ID110": {
      "attributes": [
        "a_texcoord0"
      ],
      "fragmentShader": "ID106",
      "vertexShader": "ID108"
    },
  },
```

There may be a quick fix to this in Cesium so it uses `techniques.attributes` instead.

**Comment by jbo023** [comment]
I see,  I also found the corresponding commit.
https://github.com/AnalyticalGraphicsInc/cesium/pull/4975/commits/0690a4fc2007336ef465e9988b007b0b7ddc27bb

Its a bug in the our gltf code. But before that commit cesium tried to find the attributes by using the "findVertexAttributes" function to get the attributes directly from the shader code. So the Bug in our GLTF code didn't have any impact :)

We will also change our converter to fix the b3dm files.

**Comment by lilleyse** [comment]
Yeah that's the problem area. I'm going to revert some of those changes since I don't want to break existing models.

**Comment by pjcozzi** [comment]
@lilleyse is #5008 just part of the fix here or the entire fix?

**Comment by lilleyse** [comment]
It's the entire fix. The rendering issues were resolved in #5004
